```{r echo=FALSE}
opts_chunk$set(echo=TRUE, dpi=300, fig.width=8, fig.height=5)
opts_knit$set(root.dir="/export2/home/uesu/github/readSim/example")
knit_hooks$set(htmlcap = function(before, options, envir) {
  if(!before) {
    paste('<center><p class="caption">',options$htmlcap,"</p></center>",sep="")
    }
    })
```

Selecting leaf taxa for genera without complete genomes.
====


```{r message=FALSE}
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
```

## Introduction
For genera without leaf taxa (with completed genome), 
we select a viable leaf taxon based on (1) availability of of WGS records and (2) its total sequence length.

## Exploratory analysis

For each genus belongs numerous leaf taxa. 
We investigate:
1. total number of sequences (within the genus), 
2. the number of taxa within the genus
3. the combined sequence lengths

```{r}
data=read.table("./out/readSim.0101.output.txt",sep="\t", h=T)
str(data)
#Exploratory analysis for the distribution of taxa within a given genus with relation to its total sequence length, number of sequences etc.
d2=data         %>% 
group_by(genus) %>%
summarise(numSeq=n(),
          numtaxa=length(unique(taxid)), 
          totLength=sum(as.numeric(combinedLength))) %>%
mutate(ratio = numSeq/numtaxa)
numGenus = length(unique(data$genus))
numTaxa  = length(unique(data$taxid))
```

A total of `r numGenus` genera were without completed completed genomes. 
A total of `r numTaxa` leaf taxa under the former.

```{r htmlcap = "<b>Figure:</b> Number of fastA sequences within a genus against the number of taxa within a the same genus. Annotated genus are based on the following: > 20000 sequences or >50 taxa"}
ggplot(d2, aes(y=numSeq, x=numtaxa))+
geom_point(aes(size=sqrt(totLength)))+
geom_text(data=subset(d2, numSeq > 20000| numtaxa >=50 ), aes(label=genus), color='red')+
xlab("# taxa")+ylab("# sequences")
```

## Data Processing

We choose the taxa based on the availability of WGS data and total sequence length.
```{r}
#Will have to write the script which produces this wgstaxa, taken from NCBI

#read in wgs annotation list
wgs = setNames(read.table("~/db/refseq/wgstaxa",h=F), c("taxa"))
wgsdata = subset(data, taxid %in% wgs$taxa)

#choose max length taxa
chosen = wgsdata %>% 
group_by(genus) %>% 
filter(combinedLength == max(combinedLength))
str(chosen)
```

### WGS availability 

`r sum(!unique(data$taxid) %in% unique(wgs$taxa))` taxa were removed because they have no WGS projects listed in NCBI, this resulted in `r nrow(data) - nrow(wgsdata)` sequences trimmed out of a total of `r nrow(data)`.

```{r htmlcap="<b>Figure:</b>We plot the concatenated lengths of the sequences and across genera. Each data point represents one leaf taxa. Highlighted in red are the selected taxa."}
ggplot() + 
geom_boxplot(data=wgsdata, aes(x=reorder(as.factor(genus), combinedLength, median), y=combinedLength))+
geom_jitter( data=wgsdata, aes(x=reorder(as.factor(genus), combinedLength, median), y=combinedLength))+
geom_point(data=subset(wgsdata, taxid %in% chosen$taxid), aes(x=as.factor(genus), y=combinedLength),color='red')+
theme(axis.text.x = element_text(angle=90, hjust=1))+
xlab("Genera")+ylab("Total Length of sequences in leaf taxa")
```

```{r}
write.table(chosen %>% select(-gi), file="out/readSim.0102.chosen_scaffolds.txt", sep="\t", quote=F, row.names=F, 
col.names=c("taxid","parentID","Chromosome.RefSeq", "total"))
```

```{r}
sessionInfo()
```
